sql: unskip TestTruncateWhileColumnBackfill and fix validation of check constraints

spaskob opened the pull request:
The test was previously skipped because it was timing out.  When the
timeout problem was fixed it revealed a bug with check constraints not
being re-added after truncate.

Fixes #43990.

Release note (bug fix): if a table with a check constraint is truncate
while being backfilled the check constraint may get lost.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49399)
<!-- Reviewable:end -->

spaskob commented:
@lucy-zhang  the test is still flaky with error
```
--- FAIL: TestTruncateWhileColumnBackfill (1.08s)
    schema_changer_test.go:4172: expected 1 check, found 0
FAIL
```
This is the failing check:
```
	if checks := tableDesc.AllActiveAndInactiveChecks(); len(checks) != 1 {
		t.Fatalf("expected 1 check, found %d", len(checks))
	}
```
It looks like this is a TRUNCATE bug but I am not sure. Could you take a look please?
AFAIK the schema change adds a check
```
ALTER TABLE t.public.test ADD COLUMN x DECIMAL NOT NULL DEFAULT 1.4::DECIMAL, ADD CHECK (x >= 0)
```
which should be preserved after the table is truncated.

spaskob commented:
@lucy-zhang PTAL
applied your patch

thoszhang reviewed (commented):
> Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49399)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/sql/schema_changer.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2CubeA7wHxO1-C41X:-MA2CubeA7wHxO1-C41Y:b-257vge) ([raw file](https://github.com/cockroachdb/cockroach/blob/23209a41c65cc89d20f5fde941f032dce0473972/pkg/sql/schema_changer.go#L419)):*
> ```Go
> 		// has already been removed. If there's a mutation provided, we should check
> 		// whether it actually exists on the table descriptor and exit the job if not.
> 		if sc.mutationID != sqlbase.InvalidMutationID {
> ```

Shouldn't we return here if there's a valid mutation ID with no mutation?
___
*[pkg/sql/schema_changer_test.go, line 4181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2H_ua7APYTupTPfVw:-MA2H_ua7APYTupTPfVx:b-7fo0zm) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeab02fd3817bd173d42d7b918775a01ea143a5/pkg/sql/schema_changer_test.go#L4181)):*
> ```Go
> 	//if k, x := tableDesc.Columns[0].Name, tableDesc.Columns[1].Name; k != "k" && x != "x" {
> 	//	t.Fatalf("columns %q, %q in descriptor", k, x)
> 	//}
> ```

Should this be removed entirely?
___
*[pkg/sql/sqlbase/structured.go, line 3115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2HnJW93RVxB5XL_wS:-MA2HnJX2NuvDRLSIFEF:bqvnma1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeab02fd3817bd173d42d7b918775a01ea143a5/pkg/sql/sqlbase/structured.go#L3115)):*
> ```Go
> 					desc.Checks = append(desc.Checks, &t.Constraint.Check)
> 				}
> 			case ConstraintToUpdate_FOREIGN_KEY:
> ```

Could you make the same change as above for foreign keys?


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49399)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang and @spaskob)

___
*[pkg/sql/schema_changer.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2CubeA7wHxO1-C41X:-MA2Zp2iDDlXRpRFvgoi:bfggkms) ([raw file](https://github.com/cockroachdb/cockroach/blob/23209a41c65cc89d20f5fde941f032dce0473972/pkg/sql/schema_changer.go#L419)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Shouldn't we return here if there's a valid mutation ID with no mutation?
</blockquote></details>

probably, I was trying not to change the current processing logic. Do you mind if I send a separate PR for this?


<!-- Sent from Reviewable.io -->

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49399)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/schema_changer_test.go, line 4181 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2H_ua7APYTupTPfVw:-MA2_MdP9SbxIepS7NyN:b6pvem3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeab02fd3817bd173d42d7b918775a01ea143a5/pkg/sql/schema_changer_test.go#L4181)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Should this be removed entirely?
</blockquote></details>

I forgot to uncomment these, these should still pass. I stress tested successfully.
___
*[pkg/sql/sqlbase/structured.go, line 3115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2HnJW93RVxB5XL_wS:-MA2_cEt1gGZ92aKs6Tf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeab02fd3817bd173d42d7b918775a01ea143a5/pkg/sql/sqlbase/structured.go#L3115)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Could you make the same change as above for foreign keys?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49399)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/sql/schema_changer.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2CubeA7wHxO1-C41X:-MA2hkcI0ONCJfYnaj2I:b-nxx9jo) ([raw file](https://github.com/cockroachdb/cockroach/blob/23209a41c65cc89d20f5fde941f032dce0473972/pkg/sql/schema_changer.go#L419)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

probably, I was trying not to change the current processing logic. Do you mind if I send a separate PR for this?
</blockquote></details>

Oh, I see. Yeah, I think that makes sense.
___
*[pkg/sql/sqlbase/structured.go, line 3116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2hPTbA7iQoai9pw6N:-MA2hPTbA7iQoai9pw6O:b-eoosbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a85e1ecb11714854a02bfc2349596e6b8471483d/pkg/sql/sqlbase/structured.go#L3116)):*
> ```Go
> 					desc.Checks = append(desc.Checks, &t.Constraint.Check)
> 				}
> 			case ConstraintToUpdate_FOREIGN_KEY:
> ```

Can you re-add a comment either here or at the top of `MakeMutationComplete` about how FK key backreferences need to be finalized separately by the caller, if applicable?


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
Also, I think the second commit should have a message and release note about fixing the bug.

spaskob reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49399)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/schema_changer.go, line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2CubeA7wHxO1-C41X:-MAhQ6911bwHe89GOnzQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/23209a41c65cc89d20f5fde941f032dce0473972/pkg/sql/schema_changer.go#L419)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Oh, I see. Yeah, I think that makes sense.
</blockquote></details>

Done.
___
*[pkg/sql/sqlbase/structured.go, line 3116 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49399#-MA2hPTbA7iQoai9pw6N:-MAhQ7WR9QKJnULR3grG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a85e1ecb11714854a02bfc2349596e6b8471483d/pkg/sql/sqlbase/structured.go#L3116)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Can you re-add a comment either here or at the top of `MakeMutationComplete` about how FK key backreferences need to be finalized separately by the caller, if applicable?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

thoszhang reviewed (approved):
> Reviewed 2 of 3 files at r4, 2 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49399)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

Commit:
sql: fix validation of check constraints

The test was previously skipped because it was timing out.  When the
timeout problem was fixed it revealed a bug with check constraints not
being re-added after truncate.

Fixes #43990.

Release note (bug fix): if a table with a check constraint is truncate
while being backfilled the check constraint may get lost.

spaskob commented:
bors r+

spaskob commented:
bors r+
